File: patmch/patmch_v1.java
import java.util.regex.*;
import java.io.*;

class patmch_v1 {
	public static void main(String[] args) throws Exception {
		BufferedReader stdin = new BufferedReader(new InputStreamReader(System.in));
		String l;
		if (args.length < 1) {
			System.out.println("Usage: java -cp . patmch_v1 pattern < in.file");
			return;
		}
		Pattern re = Pattern.compile(args[0]);
		while ((l = stdin.readLine()) != null) {
			Matcher m = re.matcher(l);
			if (m.find()) System.out.println(l);
		}
	}
}

File: patmch/patmch_v1.py
import re
import sys
import operator
import itertools

if len(sys.argv) != 1:
    sys.stderr.write("Usage: patmatch.py in.pattern < file.in\n")
    sys.exit(1)

strip_newline = operator.itemgetter(slice(0,-1))

def match(regex, lines_with_cr):
    search = re.compile(regex).search
    return itertools.ifilter(search, itertools.imap(strip_newline, lines_with_cr))

for line in match(sys.argv[1], sys.stdin):
    print line
